### ISSUE TITLE
AWS WAF v2 custom response body [feature request]

### ISSUE THREAD

**Issue by bbratchiv** [hashicorp/terraform-provider-aws#18754]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

You can instruct AWS WAF to send a custom HTTP response back to the client for rule actions or web ACL default actions that are set to block the request. Currently this configuration is not supported in Terraform.

### New or Affected Resource(s)

New Resource: aws_wafv2_response_body (?)
Affected Resource: aws_wafv2_web_acl, aws_wafv2_rule_group

### References
https://docs.aws.amazon.com/waf/latest/developerguide/customizing-the-response-for-blocked-requests.html

**Comment by dbgoytia** [comment]
Hey team! I'm new to this community and would like to investigate this if possible 🙂

**Comment by dcastroma** [comment]
Hi all,

At this issue we are speaking about custom response body, but at the same configuration place you can change headers and response body for block actions and add headers for allow actions.

https://docs.aws.amazon.com/waf/latest/APIReference/API_CustomResponse.html
https://docs.aws.amazon.com/waf/latest/APIReference/API_CustomResponseBody.html

**Comment by YakDriver** [comment]
@bbratchiv Thanks for submitting this feature request! It looks like there is already interest in this feature. Even if you are not interested in submitting a pull request, can you provide potential configuration for how you'd use the feature? This is often very helpful in making sure that we (or another contributor like @dbgoytia) exactly align in our thinking.

References in the AWS SDK for Go:
* `CustomResponseBody` https://docs.aws.amazon.com/sdk-for-go/api/service/wafv2/#CustomResponseBody
* `CustomResponse.CustomResponseBodyKey` https://docs.aws.amazon.com/sdk-for-go/api/service/wafv2/#CustomResponse
* `RuleGroup.CustomResponseBodies` https://docs.aws.amazon.com/sdk-for-go/api/service/wafv2/#RuleGroup ([create](https://docs.aws.amazon.com/sdk-for-go/api/service/wafv2/#CreateRuleGroupInput) [update](https://docs.aws.amazon.com/sdk-for-go/api/service/wafv2/#UpdateRuleGroupInput))
* `WebACL.CustomResponseBodies` https://docs.aws.amazon.com/sdk-for-go/api/service/wafv2/#WebACL ([create](https://docs.aws.amazon.com/sdk-for-go/api/service/wafv2/#CreateWebACLInput) [update](https://docs.aws.amazon.com/sdk-for-go/api/service/wafv2/#UpdateWebACLInput))

**Comment by andyalm** [comment]
I'm attempting to work on this [here](https://github.com/andyalm/terraform-provider-aws/tree/wafv2-custom-request-response). Am planning on submitting a PR if I can get it working. I have never contributed to a terraform provider before, and am a golang newbie, so I will likely take a bit of time to get it all working. Will accept any help or advice along the way.

**Comment by bermannoah** [comment]
Hey all, should I create an additional issue or issues for the ResponseCode and ResponseHeaders portion of the CustomResponseBody, or should I assume that this will be included? My main use-case for this feature is changing a ratelimit response from the AWS default of 403 to the standard 429.

**Comment by andyalm** [comment]
@bermannoah I am including `ResponseCode` and `ResponseHeaders` support in the PR I'm putting together #19415

**Comment by bermannoah** [comment]
@andyalm brilliant, thank you

**Comment by tx-kstav** [comment]
Hello all. With the merging of https://github.com/hashicorp/terraform-provider-aws/pull/19415 and release 3.43.0 of the provider, I am now able to add custom response code and custom response headers. Thank you everyone involved in that PR :)

However, this issue is referring to custom_response_body, which I see is not implemented in the aforementioned PR - should then this one stay open? Please correct me if I'm wrong.

**Comment by chhuang0123** [comment]
Hi all, it seems that custom_respose_body is not implemented in release 3.44 of the aws provide.
You could view the screenshot for reference:
![image](https://user-images.githubusercontent.com/1979741/121300724-cd4ab700-c929-11eb-8a57-434ce91ead1b.png)

**Comment by andyalm** [comment]
This was my fault. When I first linked my PR to this issue, I had plans on implementing all of the `custom_response` arguments. However, when I discovered that `custom_response_body_key` required an additional section to specify the body contents as well, I decided to cut that from scope. I failed to notice that the issue I had linked to specifically called out the body (there were other commenters emphasizing the need for `response_code` and `response_headers`, so I thought I was still addressing the things asked for).

Anyhow, I would be happy to put together another PR to address the custom response bodies. I'll link that here when I get it together.
